block_coursefeedback.php

```diff
@@ -69,12 +69,20 @@ public function get_content() {
         $config = get_config("block_coursefeedback");
         $renderer = $this->page->get_renderer("block_coursefeedback");
         $feedback = $DB->get_record("block_coursefeedback", array("id" => $config->active_feedback));
+        $coursestartgood = block_coursefeedbck_coursestartcheck_good($config, $this->page->course->id);
         $list = array();
 
+        // Show information banner if enabled and the coursestart is in range so a FB would be triggered.
+        if ($config->global_enable && $config->enable_infobanner && $coursestartgood
+                && has_capability("block/coursefeedback:viewanswers", $context) ) {
```

Comment:
Wenn der "Infobanner" wirklich nur dann angezeigt werden soll, wenn in dem jeweiligen Kurs auch ein Feedback eingeholt werden wird, dann sollte es m.E. eine Funktion geben, die einem beantwortet, _ob_ in Kurs mit ID `n` ein Feedback eingeholt werden wird.

Momentan gibt es zwar die [`block_coursefeedbck_coursestartcheck_good`](https://github.com/eberhardt/moodle-block_coursefeedback/blob/79b35245b525d39beebfa4f0a62d29b2f0380f72/lib.php#L977) Funktion, die auch hier verwendet wird, aber die prüft ja nicht alle Kriterien, sondern eben nur, ob der Kursbeginn den Anforderungen für ein Feedback entspricht.

So wird jetzt bspw. ja noch mit [`block_coursefeedbck_questions_exist`](https://github.com/eberhardt/moodle-block_coursefeedback/blob/79b35245b525d39beebfa4f0a62d29b2f0380f72/lib.php#L865) weiter unten geprüft, ob es auch schon valide Fragen in dem Feedback gibt. Wenn nicht, werden die anderen Notifications ja auch nicht angezeigt,.

Vielleicht ist das ja aber auch gewollt, dass das nicht voneinander abhängt. Dann ist das so auch OK. Ich muss gestehen, dass ich sowieso noch nicht ganz blicke, wann genau nun was genau angezeigt werden soll (im ursprünglichen Code). Warum wird z.B. `global_enable` nicht ein paar Zeilen weiter auch nochmal abgefragt, sondern stattdessen `active_feedback`?

Das ist aber auch genau mein Punkt: Es wäre leichter, die Logik nachzuvollziehen, wenn es deskriptiv benannte Funktionen gäbe wie bspw. `course_displays_feedback($courseid)` oder so, die die Kriterien abstrahieren. Auch z.B. für das Zusammenfassen von Capabilities sind solche Funktionen praktisch. Das hat aber wenig mit dem PR an sich zu tun, sondern mehr mit der "maintainability" insgesamt.

Reply:
Bei dem jetzigen Infobanner-Feature ist es nicht erforderlich, dass ein bestimmtes Feedback bereits angelegt ist, aktiv ist oder Fragen enthält. 
Die Voraussetzungen für das Anzeigen des Infobanners sind:

1. enable_infobanner --> das Infobanner ist aktiviert
2. $coursestratgood --> Der jeweilige Kursstart liegt nicht zu weit in der Vergangenheit
3. cap: viewanswers --> Anzeige nur für Trainer bzw. Rollen mit cap viewanser Rechten

Global enable muss tatsächlich gar nicht abgefragt werden weil es ja dafür zuständig ist den Block in die Kurse auszurollen... Ist der Block also da dann ist global enable auch aktiviert. Ist er nicht im Kurs vorhanden ist es deaktiviert. Man könnte die Abfrage höchstens als 2. Sicherheit drin lassen um sicherzugehen das Fehler beim deaktivieren von global_enable aufgefangen werden. Ich denke aber es ist nicht notwendig. Ich habe es also rausgenommen.

Nochmal extra 2 Funktionen für die If Abfragen welche entscheiden ob die eine oder andere Notifikation angezeigt wirt, anzulegen, sehe ich gerade nicht. Ich mag es in diesem Fall (die logik ist so kurz), dass ich in der If abfrage direkt sehe was überprüft wird und nicht erst in nocch eine Funktion muss um die Logig zu prüfen.

Ich habe aber versucht mit den Kommentaren noch etwas spezifischer zu werden.
d35f11b3a573ec2fe22153f599a326e860326b1c